File: 5-Problem-A-Day/131. Palindrome Partitioning.java
class Solution {
    List<List<String>> ans;
    ArrayList<String> curr;

    public List<List<String>> partition(String s) {
        ans = new ArrayList<>();
        curr = new ArrayList<>();
        backtrack(s, 0);
        return ans;
    }

    private void backtrack(String s, int l) {
        if (curr.size() > 0 && l >= s.length())
            ans.add(new ArrayList<String>(curr));
        for (int i = l; i < s.length(); i++) {
            if (isPalindrome(s, l, i)) {
                if (l == i)
                    curr.add(Character.toString(s.charAt(i)));
                else
                    curr.add(s.substring(l, i + 1));
                backtrack(s, i + 1);
                curr.remove(curr.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String str, int l, int r) {
        while (l < r) {
            if (str.charAt(l) != str.charAt(r))
                return false;
            l++;
            r--;
        }
        return true;
    }
}

File: 2021/131. Palindrome Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        int n = s.length();
        vector<vector<bool>> dp(n, vector<bool>(n , false));
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = i; j <= n; j++) {
                if (s[i] == s[j]) {
                    dp[i][j] = (i + 1 <= j - 1) ? dp[i + 1][j - 1] : true;
                } else{
                    dp[i][j] = false;
                }
            }
        }                
        
        vector<vector<string>> res;
        vector<string> curr;
        backtrack(s, 0, dp, res, curr);
        
        return res;
    }
    
    void backtrack(string& s, int start, vector<vector<bool>>& dp, vector<vector<string>>& res, vector<string>& curr) {        
        if (start >= s.length()) {
            res.push_back(curr);
        }
        string str;
        for (int i = start; i < s.length(); i++) {
            str += s[i];
            if (dp[start][i]) {
                curr.push_back(str);
                backtrack(s, i + 1, dp, res, curr);
                curr.pop_back();
            }
        }        
    }
    
    
    
};